File: packages/domain/lib/blog/domain/ports/post_repository.dart
import 'package:staretz_domain/blog/domain/entities/post.dart';
import 'package:staretz_domain/blog/domain/value_objects/post_slug.dart';
import 'package:staretz_domain/shared/pagination/page_criteria.dart';
import 'package:staretz_domain/shared/pagination/paginated.dart';

abstract class PostRepository {
  Future<List<Post>> findPreview(int limit);
  Future<Paginated<Post>> findPage(PageCriteria criteria);
  Future<Post?> findBySlug(PostSlug slug);
  Future<void> save(Post post);
  Future<void> update(PostSlug originalSlug, Post updatedPost);
  Future<void> delete(PostSlug slug);
}

File: src/modules/blog/domain/repositories/Post.repository.ts
import { PostCollection } from '../collections/Post.collection'
import { Post } from '../entities/Post.entity'
import { Slug } from '../value-objects/Slug.valueObject'

export abstract class PostRepository {
  abstract findAll(): Promise<PostCollection>
  abstract findBySlug(slug: Slug): Promise<Post | null>
  abstract save(post: Post): Promise<void>
}
